sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1794 @@
+            return x
+        elif isinstance(self.parent, Submodule):
+            coeffs_in_parent = self.parent.represent(elt)
+            parent_element = self.parent(coeffs_in_parent)
+            return self.represent(parent_element)
+        else:
+            raise ClosureFailure('Element outside ancestor chain of this module.')
+
+    def is_compat_submodule(self, other):
+        return isinstance(other, Submodule) and other.parent == self.parent
+
+    def __eq__(self, other):
+        if self.is_compat_submodule(other):
+            return other.QQ_matrix == self.QQ_matrix
+        return NotImplemented
+
+    def add(self, other, hnf=True, hnf_modulus=None):
+        """
+        Add this :py:class:`~.Submodule` to another.
+
+        Explanation
+        ===========
+
+        This represents the module generated by the union of the two modules'
+        sets of generators.
+
+        Parameters
+        ==========
+
+        other : :py:class:`~.Submodule`
+        hnf : boolean, optional (default=True)
+            If ``True``, reduce the matrix of the combined module to its
+            Hermite Normal Form.
+        hnf_modulus : :ref:`ZZ`, None, optional
+            If a positive integer is provided, use this as modulus in the
+            HNF reduction.
+
+        Returns
+        =======
+
+        :py:class:`~.Submodule`
+
+        """
+        d, e = self.denom, other.denom
+        m = ilcm(d, e)
+        a, b = m // d, m // e
+        B = (a * self.matrix).hstack(b * other.matrix)
+        if hnf:
+            B = hermite_normal_form(B, D=hnf_modulus)
+        return self.parent.submodule_from_matrix(B, denom=m)
+
+    def __add__(self, other):
+        if self.is_compat_submodule(other):
+            return self.add(other)
+        return NotImplemented
+
+    __radd__ = __add__
+
+    def mul(self, other, hnf=True, hnf_modulus=None):
+        """
+        Multiply this :py:class:`~.Submodule` by a rational number, a
+        :py:class:`~.ModuleElement`, or another :py:class:`~.Submodule`.
+
+        Explanation
+        ===========
+
+        To multiply by a rational number or :py:class:`~.ModuleElement` means
+        to form the submodule whose generators are the products of this
+        quantity with all the generators of the present submodule.
+
+        To multiply by another :py:class:`~.Submodule` means to form the
+        submodule whose generators are all the products of one generator from
+        the one submodule, and one generator from the other.
+
+        Parameters
+        ==========
+
+        other : int, :ref:`ZZ`, :ref:`QQ`, :py:class:`~.ModuleElement`, :py:class:`~.Submodule`
+        hnf : boolean, optional (default=True)
+            If ``True``, reduce the matrix of the product module to its
+            Hermite Normal Form.
+        hnf_modulus : :ref:`ZZ`, None, optional
+            If a positive integer is provided, use this as modulus in the
+            HNF reduction.
```

Comment:
This is explained in the `hermite_normal_form` function docstring so a clickable link from here to there like "see the explanation of this parameter at :py:func:..." would be good or at least a "See also" here.